<a href="https://colab.research.google.com/github/bestjin98/python_data_basic/blob/main/pandas_exercise_ipynb%EC%9D%98_%EC%82%AC%EB%B3%B8.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

np.random.seed(0)

# 과제 1
## Q1. 


<div class="alert alert-success">
[문제1]     
    
insurance.csv'파일 읽어서 df에 할당하고, 상위 5개 항목 출력하기
</div>

------------

[힌트]

pd.read_csv()

In [2]:
df = pd.read_csv('https://raw.githubusercontent.com/zzhining/ml_basic/main/dataset/insurance.csv')
df.head()

age     sex     bmi  children smoker     region      charges
0   19  female  27.900         0    yes  southwest  16884.92400
1   18    male  33.770         1     no  southeast   1725.55230
2   28    male  33.000         3     no  southeast   4449.46200
3   33    male  22.705         0     no  northwest  21984.47061
4   32    male  28.880         0     no  northwest   3866.85520

## Q2. 


<div class="alert alert-success">
[문제2]     
    
df의 요약(평균, 4분위수, 전체 갯수 등)정보 확인하기
</div>

------------

In [3]:
df.describe()

age          bmi     children       charges
count  1338.000000  1338.000000  1338.000000   1338.000000
mean     39.207025    30.663397     1.094918  13270.422265
std      14.049960     6.098187     1.205493  12110.011237
min      18.000000    15.960000     0.000000   1121.873900
25%      27.000000    26.296250     0.000000   4740.287150
50%      39.000000    30.400000     1.000000   9382.033000
75%      51.000000    34.693750     2.000000  16639.912515
max      64.000000    53.130000     5.000000  63770.428010

## Q3. 


<div class="alert alert-success">
[문제3]     
    
신규 컬럼(double bmi)을 생성하고, bmi의 2배가 되는 값을 할당하기
</div>

------------

In [4]:
df['double bmi'] = df['bmi']*2

## Q4. 


<div class="alert alert-success">
[문제4]     
    
신규 컬럼(debt)을 생성하고, 0으로 채우기
</div>

------------

In [5]:
df['debt'] = 0

## Q5. 


<div class="alert alert-success">
[문제5]     
    
debt에 0부터 값이 증가하도록 채우기
</div>

------------

In [6]:
df['debt']=range(len(df))

In [7]:
df.head()

age     sex     bmi  children smoker     region      charges  double bmi  \
0   19  female  27.900         0    yes  southwest  16884.92400       55.80   
1   18    male  33.770         1     no  southeast   1725.55230       67.54   
2   28    male  33.000         3     no  southeast   4449.46200       66.00   
3   33    male  22.705         0     no  northwest  21984.47061       45.41   
4   32    male  28.880         0     no  northwest   3866.85520       57.76   

   debt  
0     0  
1     1  
2     2  
3     3  
4     4

## Q6. 


<div class="alert alert-success">
[문제6]     
    
1번 인덱스의 정보 출력
</div>

------------

[힌트]
iloc



age                  18  
sex                male  
bmi               33.77  
children              1  
smoker               no  
region        southeast  
charges         1725.55  
double bmi        67.54  
debt                  1  

In [8]:
df.iloc[1]

age                  18
sex                male
bmi               33.77
children              1
smoker               no
region        southeast
charges       1725.5523
double bmi        67.54
debt                  1
Name: 1, dtype: object

## Q7. 


<div class="alert alert-success">
[문제7]     
    
'double bmi'열 삭제하기
</div>

In [9]:
df.drop(['double bmi'], axis = 1, inplace = True)

## Q8. 


<div class="alert alert-success">
[문제8]     
    
 'age'열, 'sex'열만 출력하기
</div>

In [10]:
df[['age', 'sex']]

age     sex
0      19  female
1      18    male
2      28    male
3      33    male
4      32    male
...   ...     ...
1333   50    male
1334   18  female
1335   18  female
1336   21  female
1337   61  female

[1338 rows x 2 columns]

# 과제2
## Q9.
<div class="alert alert-success">
[문제9]     
    
나이를 1살씩 증가시키기
</div>

In [12]:
df['age']=df['age']+1
df.head()

age     sex     bmi  children smoker     region      charges  debt
0   21  female  27.900         0    yes  southwest  16884.92400     0
1   20    male  33.770         1     no  southeast   1725.55230     1
2   30    male  33.000         3     no  southeast   4449.46200     2
3   35    male  22.705         0     no  northwest  21984.47061     3
4   34    male  28.880         0     no  northwest   3866.85520     4

## Q10.
<div class="alert alert-success">
[문제10]     
    
charges의 표시형식을 소수점 아래 2자리로 바꾸기
</div>

[힌트]

함수 생성, 적용(apply/map)

In [13]:
df['charges'] = round(df['charges'],2)

In [14]:
df['charges'].round(2)

0       16884.92
1        1725.55
2        4449.46
3       21984.47
4        3866.86
          ...   
1333    10600.55
1334     2205.98
1335     1629.83
1336     2007.94
1337    29141.36
Name: charges, Length: 1338, dtype: float64

In [15]:
def sosu(data):
    return format(data,".2f")

df['charges'].apply(sosu)

0       16884.92
1        1725.55
2        4449.46
3       21984.47
4        3866.86
          ...   
1333    10600.55
1334     2205.98
1335     1629.83
1336     2007.94
1337    29141.36
Name: charges, Length: 1338, dtype: object

In [16]:
f = lambda x: '%.2f' % x
df['charges'].apply(f)

0       16884.92
1        1725.55
2        4449.46
3       21984.47
4        3866.86
          ...   
1333    10600.55
1334     2205.98
1335     1629.83
1336     2007.94
1337    29141.36
Name: charges, Length: 1338, dtype: object

## Q11.
<div class="alert alert-success">
[문제11]     
    
컬럼 정렬(알파벳 순서대로) : age, bmi, charges... 순으로 정렬
</div>

[Hint]
sort_index()


In [17]:
df = df.sort_index(axis=1)

In [18]:
df = df.sort_index(axis=1)

In [19]:
df.sort_index(axis=1)

age     bmi   charges  children  debt     region     sex smoker
0      21  27.900  16884.92         0     0  southwest  female    yes
1      20  33.770   1725.55         1     1  southeast    male     no
2      30  33.000   4449.46         3     2  southeast    male     no
3      35  22.705  21984.47         0     3  northwest    male     no
4      34  28.880   3866.86         0     4  northwest    male     no
...   ...     ...       ...       ...   ...        ...     ...    ...
1333   52  30.970  10600.55         3  1333  northwest    male     no
1334   20  31.920   2205.98         0  1334  northeast  female     no
1335   20  36.850   1629.83         0  1335  southeast  female     no
1336   23  25.800   2007.94         0  1336  southwest  female     no
1337   63  29.070  29141.36         0  1337  northwest  female    yes

[1338 rows x 8 columns]

## Q12.
<div class="alert alert-success">
[문제12]     
    
값 정렬(age) : 나이 순으로 정렬
</div>

[Hint]
sort_values


In [20]:
df.sort_values(by='age')

age     bmi   charges  children  debt     region     sex smoker
1248   20  39.820   1633.96         0  1248  southeast  female     no
482    20  31.350   1622.19         0   482  southeast  female     no
492    20  25.080   2196.47         0   492  northeast  female     no
525    20  33.880  11482.63         0   525  southeast  female     no
529    20  25.460   1708.00         0   529  northeast    male     no
...   ...     ...       ...       ...   ...        ...     ...    ...
398    66  25.600  14988.43         2   398  southwest    male     no
335    66  34.500  13822.80         0   335  southwest    male     no
378    66  30.115  16455.71         3   378  northwest  female     no
1265   66  23.760  26926.51         0  1265  southeast    male    yes
635    66  38.190  14410.93         0   635  northeast    male     no

[1338 rows x 8 columns]

## Q13.
<div class="alert alert-success">
[문제13]     
    
나이('age') 평균구하기
</div>

[Hint]
mean

In [23]:
round(df['age'].mean(), 2)

41.21

In [24]:
round(df['age'].sum() / len(df), 2)

41.21

## Q14.
<div class="alert alert-success">
[문제14]     
    
보험료('charges')와 가장 상관관계가 높은 변수는?
</div>

[Hint]
 - 공분산: df.cov()
 - 상관계수: df.corr()

In [25]:
df.head()

age     bmi   charges  children  debt     region     sex smoker
0   21  27.900  16884.92         0     0  southwest  female    yes
1   20  33.770   1725.55         1     1  southeast    male     no
2   30  33.000   4449.46         3     2  southeast    male     no
3   35  22.705  21984.47         0     3  northwest    male     no
4   34  28.880   3866.86         0     4  northwest    male     no

In [26]:
df.cov()

age           bmi       charges    children           debt
age         197.401387      9.362337  5.087480e+04    0.719303    -170.829095
bmi           9.362337     37.187884  1.464730e+04    0.093795     -85.224706
charges   50874.801787  14647.304865  1.466524e+08  992.674264  -15781.315299
children      0.719303      0.093795  9.926743e+02    1.453213      11.747569
debt       -170.829095    -85.224706 -1.578132e+04   11.747569  149298.500000

In [27]:
df.corr()

age       bmi   charges  children      debt
age       1.000000  0.109272  0.299008  0.042469 -0.031467
bmi       0.109272  1.000000  0.198341  0.012759 -0.036169
charges   0.299008  0.198341  1.000000  0.067998 -0.003373
children  0.042469  0.012759  0.067998  1.000000  0.025221
debt     -0.031467 -0.036169 -0.003373  0.025221  1.000000

# 과제 3
## Q15.
<div class="alert alert-success">
[문제15]     
    
 남자, 여자가 각각 몇명있는지 구하기
</div>

[Hint]
pd.value_counts


In [28]:
df['sex'].value_counts()

male      676
female    662
Name: sex, dtype: int64

In [29]:
pd.value_counts(df['sex'])

male      676
female    662
Name: sex, dtype: int64

In [30]:
pd.value_counts(df.sex)

male      676
female    662
Name: sex, dtype: int64

## Q16.
<div class="alert alert-success">
[문제16]     
    
자녀(children)가 3명인 데이터 출력하기
</div>

In [31]:
df[df['children'] == 3]

age     bmi   charges  children  debt     region     sex smoker
2      30  33.000   4449.46         3     2  southeast    male     no
7      39  27.740   7281.51         3     7  northwest  female     no
25     61  27.720  14001.13         3    25  southeast  female     no
36     64  32.965  15612.19         3    36  northwest  female     no
54     42  28.690   8059.68         3    54  northwest  female     no
...   ...     ...       ...       ...   ...        ...     ...    ...
1301   64  30.875  46718.16         3  1301  northwest    male    yes
1314   32  23.655  18765.88         3  1314  northwest  female    yes
1320   33  31.065   5425.02         3  1320  northwest    male     no
1332   54  44.700  11411.68         3  1332  southwest  female     no
1333   52  30.970  10600.55         3  1333  northwest    male     no

[157 rows x 8 columns]

## Q16.
<div class="alert alert-success">
[문제16]     
    
자녀(children)가  5명 이상인 데이터 출력하기
</div>

In [32]:
df[df['children'] >= 5]

age     bmi   charges  children  debt     region     sex smoker
32     21  28.600   4687.80         5    32  southwest  female     no
71     33  28.500   6799.46         5    71  northeast    male     no
166    22  37.000   4830.63         5   166  southwest  female     no
413    27  23.900   5080.10         5   413  southwest    male     no
425    47  24.310   9788.87         5   425  southeast    male     no
438    54  46.750  12592.53         5   438  southeast  female     no
568    51  31.900  11552.90         5   568  southwest  female     no
640    35  42.400   6666.24         5   640  southwest    male     no
877    35  33.440   6653.79         5   877  southeast    male     no
932    48  25.800  10096.97         5   932  southwest    male     no
937    41  24.225   8965.80         5   937  northwest  female     no
969    41  34.320   8596.83         5   969  southeast  female     no
984    22  30.115   4915.06         5   984  northeast    male     no
1085   41  18.300  19023.26         5  1085  southwest  female    yes
1116   43  29.640   9222.40         5  1116  northeast    male     no
1130   41  23.870   8582.30         5  1130  southeast  female     no
1245   30  24.300   5615.37         5  1245  southwest    male     no
1272   45  25.520  14478.33         5  1272  southeast    male     no

## Q17.
<div class="alert alert-success">
[문제17]     
    
 bmi가  17 미만인 데이터 출력하기
</div>

In [33]:
df[df['bmi'] < 17]

age     bmi  charges  children  debt     region     sex smoker
172    20  15.960  1694.80         0   172  northeast    male     no
428    23  16.815  3167.46         1   428  northeast  female     no
1226   40  16.815  6640.54         2  1226  northeast    male     no

# 과제 4 
## Q18.
<div class="alert alert-success">
[문제18]     
    
성별에 따른 보험료 평균
</div>

In [34]:
df['charges'].groupby(df['sex']).mean()

sex
female    12569.578716
male      13956.751331
Name: charges, dtype: float64

In [35]:
df.groupby('sex')['charges'].mean()

sex
female    12569.578716
male      13956.751331
Name: charges, dtype: float64

In [36]:
df.groupby(['sex']).mean()['charges']

sex
female    12569.578716
male      13956.751331
Name: charges, dtype: float64

## Q19.
<div class="alert alert-success">
[문제19]     
    
흡연여부에 따른 보험료 평균
</div>

In [37]:
df.columns

Index(['age', 'bmi', 'charges', 'children', 'debt', 'region', 'sex', 'smoker'], dtype='object')

In [38]:
df['charges'].groupby(df['smoker']).mean()

smoker
no      8434.268299
yes    32050.231898
Name: charges, dtype: float64

## Q20.
<div class="alert alert-success">
[문제20]     
    
지역별 보험료 평균
</div>

In [39]:
df['charges'].groupby(df['region']).mean()

region
northeast    13406.384599
northwest    12417.575169
southeast    14735.411484
southwest    12346.937508
Name: charges, dtype: float64

## Q21.
<div class="alert alert-success">
[문제21]     
    
흡연여부 + 지역별 데이터 평균
</div>

In [40]:
df.groupby(['smoker','region']).mean()

age        bmi       charges  children        debt
smoker region                                                             
no     northeast  41.536965  29.332082   9165.531829  1.046693  660.669261
       northwest  41.168539  29.212678   8556.463558  1.119850  679.161049
       southeast  40.673993  33.442418   8032.216264  1.065934  650.109890
       southwest  42.183521  30.507865   8019.284569  1.127341  680.104869
yes    northeast  40.238806  28.565224  29673.536269  1.044776  682.328358
       northwest  41.327586  29.140431  30192.002759  1.275862  699.534483
       southeast  41.736264  33.096703  34844.997143  1.000000  665.714286
       southwest  38.103448  31.005172  32269.063966  1.206897  644.620690

## Q22.
<div class="alert alert-success">
[문제22]     
    
 흡연여부 + 지역별 데이터 평균에서 'charges' 만 츨력하기
</div>

In [41]:
df.groupby(['smoker','region'])['charges'].mean()

smoker  region   
no      northeast     9165.531829
        northwest     8556.463558
        southeast     8032.216264
        southwest     8019.284569
yes     northeast    29673.536269
        northwest    30192.002759
        southeast    34844.997143
        southwest    32269.063966
Name: charges, dtype: float64

## Q23.
<div class="alert alert-success">
[문제23]     
    
 흡연여부 + 지역별 데이터 크기(size)
</div>

In [42]:
df.groupby(['smoker','region']).count()

age  bmi  charges  children  debt  sex
smoker region                                           
no     northeast  257  257      257       257   257  257
       northwest  267  267      267       267   267  267
       southeast  273  273      273       273   273  273
       southwest  267  267      267       267   267  267
yes    northeast   67   67       67        67    67   67
       northwest   58   58       58        58    58   58
       southeast   91   91       91        91    91   91
       southwest   58   58       58        58    58   58

## Q24.
<div class="alert alert-success">
[문제24]     
    
smoker 여부에 따라, group_val 컬럼에 '흡연자', '비흡연자' 표시하기
</div>

In [43]:
df['group_val'] = '흡연자'

In [44]:
df.drop('group_val', axis = 1, inplace=True)

In [45]:
df2 = pd.DataFrame({'group_val' : ['흡연자', '비흡연자']}, index =['yes', 'no'])
df2

group_val
yes       흡연자
no       비흡연자

In [46]:
pd.merge(df, df2, left_on='smoker', right_index = True, how = 'outer' )

age    bmi   charges  children  debt     region     sex smoker group_val
0      21  27.90  16884.92         0     0  southwest  female    yes       흡연자
11     64  26.29  27808.73         0    11  southeast  female    yes       흡연자
14     29  42.13  39611.76         0    14  southeast    male    yes       흡연자
19     32  35.30  36837.47         0    19  southwest    male    yes       흡연자
23     36  31.92  37701.88         1    23  northeast  female    yes       흡연자
...   ...    ...       ...       ...   ...        ...     ...    ...       ...
1332   54  44.70  11411.68         3  1332  southwest  female     no      비흡연자
1333   52  30.97  10600.55         3  1333  northwest    male     no      비흡연자
1334   20  31.92   2205.98         0  1334  northeast  female     no      비흡연자
1335   20  36.85   1629.83         0  1335  southeast  female     no      비흡연자
1336   23  25.80   2007.94         0  1336  southwest  female     no      비흡연자

[1338 rows x 9 columns]

In [47]:
def smok(sm):
    if sm=="yes":
        return("흡연자")
    else:
        return("비흡연자")

df['group_val']=df['smoker'].apply(smok)
df

age     bmi   charges  children  debt     region     sex smoker  \
0      21  27.900  16884.92         0     0  southwest  female    yes   
1      20  33.770   1725.55         1     1  southeast    male     no   
2      30  33.000   4449.46         3     2  southeast    male     no   
3      35  22.705  21984.47         0     3  northwest    male     no   
4      34  28.880   3866.86         0     4  northwest    male     no   
...   ...     ...       ...       ...   ...        ...     ...    ...   
1333   52  30.970  10600.55         3  1333  northwest    male     no   
1334   20  31.920   2205.98         0  1334  northeast  female     no   
1335   20  36.850   1629.83         0  1335  southeast  female     no   
1336   23  25.800   2007.94         0  1336  southwest  female     no   
1337   63  29.070  29141.36         0  1337  northwest  female    yes   

     group_val  
0          흡연자  
1         비흡연자  
2         비흡연자  
3         비흡연자  
4         비흡연자  
...        ...  
1333      비흡연자  
1334      비흡연자  
1335      비흡연자  
1336      비흡연자  
1337       흡연자  

[1338 rows x 9 columns]

In [48]:
df['group_val'] = df.replace(['yes', 'no'], ['흡연자', '비흡연자'])['smoker']

In [49]:
data={'smoker':["yes", "no"],
     'group_val':["흡연자", "비흡연자"]}
df1=pd.DataFrame(data)
pd.merge(df, df1, how='left')

age     bmi   charges  children  debt     region     sex smoker  \
0      21  27.900  16884.92         0     0  southwest  female    yes   
1      20  33.770   1725.55         1     1  southeast    male     no   
2      30  33.000   4449.46         3     2  southeast    male     no   
3      35  22.705  21984.47         0     3  northwest    male     no   
4      34  28.880   3866.86         0     4  northwest    male     no   
...   ...     ...       ...       ...   ...        ...     ...    ...   
1333   52  30.970  10600.55         3  1333  northwest    male     no   
1334   20  31.920   2205.98         0  1334  northeast  female     no   
1335   20  36.850   1629.83         0  1335  southeast  female     no   
1336   23  25.800   2007.94         0  1336  southwest  female     no   
1337   63  29.070  29141.36         0  1337  northwest  female    yes   

     group_val  
0          흡연자  
1         비흡연자  
2         비흡연자  
3         비흡연자  
4         비흡연자  
...        ...  
1333      비흡연자  
1334      비흡연자  
1335      비흡연자  
1336      비흡연자  
1337       흡연자  

[1338 rows x 9 columns]